### PR TITLE
Create bytes array bitmap library

### PR DESCRIPTION
**Pull request by ChaoticWalrus**
Simple utility library for converting between bytes arrays and bitmaps.
Currently experimental; function signatures may change.
Further optimizations are 100% possible.

### CODE REVIEW

File: src/contracts/libraries/BytesArrayBitmaps.sol [hunk 1]
```diff
@@ -0,0 +1,139 @@
...
+ * @author Layr Labs, Inc.
+ */
+library BytesArrayBitmaps {
+    /**
+     * @notice Converts an array of bytes into a bitmap.
+     * @param bytesArray The array of bytes to convert/compress into a bitmap.
+     * @return The resulting bitmap.
+     * @dev Each byte in the input is processed as indicating a single bit to flip in the bitmap
+     */
+    function bytesArrayToBitmap(bytes calldata bytesArray) internal pure returns (uint256) {
```
Comment (gpsanant) [inline src/contracts/libraries/BytesArrayBitmaps.sol:16]: Didn't understand this was what the non-ordered version would look like. Definitely cleaner, dump the ordered version :)

Comment (ChaoticWalrus) [reply src/contracts/libraries/BytesArrayBitmaps.sol:16]: wait until you see the Yul version

File: src/contracts/libraries/BytesArrayBitmaps.sol [hunk 2]
```diff
@@ -0,0 +1,197 @@
...
+
+        // loop through each byte in the array to construct the bitmap
+        for (uint256 i = 1; i < orderedBytesArray.length; ++i) {
+            // check that the entry is *strictly greater than* the previous entry. enforces both ordering and non-duplication in the array
+            require(uint256(uint8(orderedBytesArray[i])) > uint256(uint8(singleByte)),
+                "BytesArrayBitmaps.orderedBytesArrayToBitmap: orderedBytesArray is not ordered");
+            // pull the next byte out of the array
+            singleByte = orderedBytesArray[i];
+            // construct a single-bit mask from the numerical value of the byte
+            bitMask = uint256(1 << uint8(singleByte));
```
Comment (gpsanant) [inline src/contracts/libraries/BytesArrayBitmaps.sol:74]: could we do the same check you wrote below here? just 
```solidity
bitMask = uint256(1 << uint8(orderedBytesArray[i]));
require(bitmask > bitmap)
```

**Comment by gpsanant** [comment]
If https://github.com/Layr-Labs/eigenlayer-contracts/pull/49#discussion_r1207016222 is true, then my recommendation is we simply implement that and we can use the YUL in the registries.

File: src/contracts/libraries/BytesArrayBitmaps.sol [hunk 3]
```diff
@@ -0,0 +1,197 @@
...
+    /**
+     * @notice Utility function for checking if a bytes array is strictly ordered, in ascending order.
+     * @param bytesArray the bytes array of interest
+     * @return Returns 'true' if the array is ordered in strictly ascending order, and 'false' otherwise.
+     * @dev This function returns 'true' for the edge case of the `bytesArray` having zero length.
+     *      It also returns 'false' early for arrays with length in excess of 255 (i.e. so long that they cannot be strictly ordered)
+     */
+    function isArrayStrictlyAscendingOrdered(bytes calldata bytesArray) internal pure returns (bool) {
+        // return 'false' early for too-long (i.e. unorderable) arrays
+        if (bytesArray.length > 255) {
```
[the remark this answers is not in the document]
Comment (ChaoticWalrus) [reply src/contracts/libraries/BytesArrayBitmaps.sol:148]: ahh good point, you're right, will fix

File: src/contracts/libraries/BytesArrayBitmaps.sol (hunk 2 above)
Comment (ChaoticWalrus) [reply src/contracts/libraries/BytesArrayBitmaps.sol:74]: yes, I'll change that, it's definitely more efficient

File: src/contracts/libraries/BytesArrayBitmaps.sol (hunk 3 above)
[the remark this answers is not in the document]
Comment (ChaoticWalrus) [reply src/contracts/libraries/BytesArrayBitmaps.sol]: fixed in https://github.com/Layr-Labs/eigenlayer-contracts/pull/49/commits/857740bc449ad16c4f5e1cbca0fdc5ddfbdd9168

File: src/contracts/libraries/BytesArrayBitmaps.sol (hunk 2 above)
Comment (ChaoticWalrus) [reply src/contracts/libraries/BytesArrayBitmaps.sol]: done in https://github.com/Layr-Labs/eigenlayer-contracts/pull/49/commits/857740bc449ad16c4f5e1cbca0fdc5ddfbdd9168

File: src/test/unit/BytesArrayBitmapsUnit.t.sol [hunk 4]
```diff
@@ -0,0 +1,122 @@
...
+    Vm cheats = Vm(HEVM_ADDRESS);
+
+    BytesArrayBitmapsWrapper public bytesArrayBitmapsWrapper;
+
+    function setUp() public {
+        bytesArrayBitmapsWrapper = new BytesArrayBitmapsWrapper();
+    }
+
+    // ensure that the bitmap encoding of a single uint8 (i.e. a single byte) matches the expected output
+    function testSingleByteEncoding(uint8 fuzzedNumber) public view {
```
[the remark this answers is not in the document]
Comment (ChaoticWalrus) [reply src/test/unit/BytesArrayBitmapsUnit.t.sol:19]: good point, will add

**Comment by ChaoticWalrus** [comment]
> If [#49 (comment)](https://github.com/Layr-Labs/eigenlayer-contracts/pull/49#discussion_r1207016222) is true, then my recommendation is we simply implement that and we can use the YUL in the registries.

It works but is still less efficient 😄

File: src/test/unit/BytesArrayBitmapsUnit.t.sol (hunk 4 above)
[the remark this answers is not in the document]
Comment (ChaoticWalrus) [reply src/test/unit/BytesArrayBitmapsUnit.t.sol:19]: actually, this raises an interesting point -- should the bitmap for an empty array be a uint(0) -- as I have it now -- OR should we revert?

[the remark this answers is not in the document]
Comment (ChaoticWalrus) [reply src/test/unit/BytesArrayBitmapsUnit.t.sol:19]: agreed offline that returning 0 seems more appropriate for a library. more specific uses may wish to revert in the case of an empty array/bitmap.

[the remark this answers is not in the document]
Comment (ChaoticWalrus) [reply src/test/unit/BytesArrayBitmapsUnit.t.sol:26]: test added here https://github.com/Layr-Labs/eigenlayer-contracts/pull/49/commits/56cd21d7c47dedab674cb94fb331560d1a502f79

**Review by gpsanant — APPROVED** [review]
looks good

**Pull request merged**